main/src/addins/MonoDevelop.TextEditor/MonoDevelop.TextEditor/TextViewDisplayBinding.cs

```diff
@@ -45,56 +45,47 @@ protected override IEnumerable<DocumentControllerDescription> GetSupportedContro
 				yield break;
 			}
 
-			if (modelDescriptor.FilePath == null || !(IsSupportedFileExtension (modelDescriptor.FilePath) || IsSupportedDesignerFileName (modelDescriptor.FilePath, modelDescriptor.Owner))) {
-				yield break;
-			}
-
-			bool supported = false;
+			var nodes = Mono.Addins.AddinManager.GetExtensionNodes<SupportedFileTypeExtensionNode> ("/MonoDevelop/TextEditor/SupportedFileTypes");
 
-			if (modelDescriptor.FilePath != null)
-				supported = IdeServices.DesktopService.GetFileIsText (modelDescriptor.FilePath, modelDescriptor.MimeType);
-
-			if (!supported && !string.IsNullOrEmpty (modelDescriptor.MimeType))
-				supported = IdeServices.DesktopService.GetMimeTypeIsText (modelDescriptor.MimeType);
+			bool supported =
+				(
+					modelDescriptor.FilePath.IsNotNull
+					&& IdeServices.DesktopService.GetFileIsText (modelDescriptor.FilePath, modelDescriptor.MimeType)
+					&& nodes.Any (n => ExtensionMatch (n) && BuildActionAndFeatureFlagMatch (n))
+				) || (
+					!string.IsNullOrEmpty (modelDescriptor.MimeType)
+					&& IdeServices.DesktopService.GetMimeTypeIsText (modelDescriptor.MimeType)
+					&& nodes.Any (n => MimeMatch (n) && BuildActionAndFeatureFlagMatch (n))
+				);
 
 			if (supported)
 				yield return new DocumentControllerDescription (GettextCatalog.GetString ("New Source Code Editor"), true, DocumentControllerRole.Source);
-		}
 
-		static HashSet<string> supportedFileExtensions = new HashSet<string> (StringComparer.OrdinalIgnoreCase) {
-			".cs",
-			".csx"
-			//".cshtml",
-			//".css",
-			//".html",
-			//".js",
-			//".json",
-			//".ts"
-		};
+			bool ExtensionMatch (SupportedFileTypeExtensionNode node) =>
+				node.Extensions != null
+				&& node.Extensions.Any (modelDescriptor.FilePath.HasExtension);
```

Comment:
Prefer explicit lambda arguments to method group. Method Groups have poor code generation on the compiler side.

```suggestion
				&& node.Extensions.Any (ext => modelDescriptor.FilePath.HasExtension (ext));
```

Reply:
Ugh, I keep forgetting about that one. Can we fix the compiler already?

Reply:
https://github.com/dotnet/roslyn/pull/6642